File: medium/3876-construct-uniform-parity-array-ii/solution.py
from typing import List


class Solution:
    def canConstruct(self, nums1: List[int]) -> bool:
        """
        3876. Construct Uniform Parity Array II
        Time: O(n)
        Space: O(1)
        """
        INF = float("inf")
        min_odd = INF
        min_even = INF
        odd_count = 0

        for x in nums1:
            if x % 2:
                odd_count += 1
                if x < min_odd:
                    min_odd = x
            elif x < min_even:
                min_even = x

        even_count = len(nums1) - odd_count

        # The smallest odd can never flip (nothing odd is below it),
        # so any odd element blocks an all-even result.
        all_even = odd_count == 0
        # Every even must flip, which needs an odd strictly below it.
        all_odd = even_count == 0 or (odd_count >= 1 and min_odd < min_even)

        return all_even or all_odd

File: medium/3876-construct-uniform-parity-array-ii/solution.java
/**
 * 3876. Construct Uniform Parity Array II
 * Time: O(n)
 * Space: O(1)
 */
class Solution {
    public boolean canConstruct(int[] nums1) {
        long minOdd = Long.MAX_VALUE;
        long minEven = Long.MAX_VALUE;
        int oddCount = 0;

        for (int x : nums1) {
            if (x % 2 != 0) {
                oddCount++;
                if (x < minOdd) {
                    minOdd = x;
                }
            } else if (x < minEven) {
                minEven = x;
            }
        }

        int evenCount = nums1.length - oddCount;

        // The smallest odd can never flip (nothing odd is below it),
        // so any odd element blocks an all-even result.
        boolean allEven = oddCount == 0;
        // Every even must flip, which needs an odd strictly below it.
        boolean allOdd = evenCount == 0 || (oddCount >= 1 && minOdd < minEven);

        return allEven || allOdd;
    }
}
